### PR TITLE
Mixed Precision Support

### PR DESCRIPTION
**Pull request by ycheng223**
Adds mixed precision support via JMP library. Per JMP policies, converts model parameters (param), forward/back pass computations (compute), and/or  activations+outputs (output) to either full (32 bit) or half (16 bit) precision. 

The main feature is the new MixedPrecisionMixin class in xax/task/mixins/mixed_precision.py which overrides some methods from supervised.py when called (get_model, get_output_and_loss, update, val_step) . The mixin also includes a dynamic loss scaling method to handle potential gradient underflow caused by FP16.

Other adjustments include importing mixed_precision_mixin to task/mixins/__init.py__ and updating requirements.txt. I also ran into a circular import error in single_process.py when running scripts/update_api.py. This was fixed by making get_num_gpus a lazy import in def configure_gpu_devices. 

Refer to the mnist_mp.py or test_train_e2e_mp.py examples for sample mixed precision implementation. The main additions needed to add mixed precision and configure policy is below:

class Config(**xax.MixedPrecisionConfig**, xax.SupervisedConfig):
    batch_size: int = xax.field(128, help="The size of a minibatch")
    learning_rate: float = xax.field(1e-3, help="The learning rate")
    hidden_dim: int = xax.field(512, help="Hidden layer dimension")
    num_hidden_layers: int = xax.field(2, help="Number of hidden layers")

    # Mixed precision configuration
    **precision_policy: str = xax.field("half_param", help="Mixed precision policy")**
    

class MnistClassification(**xax.MixedPrecisionMixin[Config]**, xax.SupervisedTask[Config]):
    def get_model(self, params: xax.InitParams) -> Model:
        return Model(
            self.config.num_hidden_layers,
            self.config.hidden_dim,
            key=params.key,
        )
        
**Policy Configurations**
            "full": jmp.get_policy("params=float32,compute=float32,output=float32"),
            "half_param": jmp.get_policy("params=float16,compute=float16,output=float32"),
            "half_compute": jmp.get_policy("params=float32,compute=float16,output=float32"), 
            "half_output": jmp.get_policy("params=float32,compute=float32,output=float16"),
            "half_front": jmp.get_policy("params=float16,compute=float16,output=float32"),
            "half_back": jmp.get_policy("params=float32,compute=float16,output=float16"),
            "half_total": jmp.get_policy("params=float16,compute=float16,output=float16")
            
 
I also created a quick and dirty benchmarking/comparison tool for measuring the performance of different policies using the a modified version (MP-enabled) of the MNIST example. This code can benchmark any policy, compare the speed and memory usage of any two policies, or perform a comparison of all policies. The code can be found here:
 
 https://github.com/ycheng223/xax_mp_benchmarking/tree/master
 
[Screencast from 09-18-2025 07:16:46 PM.webm](https://github.com/user-attachments/assets/36500886-cdaf-4f82-b009-90cc85996166)

### CODE REVIEW

File: xax/requirements.txt [hunk 1]
```diff
@@ -10,6 +10,10 @@ jax
 jaxtyping
 optax
 orbax-checkpoint
+<<<<<<< HEAD
+jmp
+=======
+>>>>>>> 7d92e76bd7ef08c9d19642c03bd7679375bf6689
```
Comment (codekansas) [inline xax/requirements.txt:16]: need to fix the merge errors

File: xax/task/mixins/mixed_precision.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            "half_param": jmp.get_policy("params=float16,compute=float16,output=float32"),
+            "half_compute": jmp.get_policy("params=float32,compute=float16,output=float32"), 
+            "half_output": jmp.get_policy("params=float32,compute=float32,output=float16"),
+            "half_front": jmp.get_policy("params=float16,compute=float16,output=float32"),
+            "half_back": jmp.get_policy("params=float32,compute=float16,output=float16"),
+            "half_total": jmp.get_policy("params=float16,compute=float16,output=float16")
+        }
+        
+        if self.config.precision_policy not in policies:
+            logger.warning(f"Unknown precision policy '{self.config.precision_policy}', falling back to 'full'")
```
Comment (codekansas) [inline xax/task/mixins/mixed_precision.py:74]: prefer to raise an exception

File: xax/task/mixins/mixed_precision.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        # Cast outputs to output precision
+        output_casted = self.mp_policy.cast_to_output(output)
+        metrics_casted = {k: self.mp_policy.cast_to_output(v) for k, v in metrics.items()}
+        
+        return output_casted, FrozenDict(metrics_casted)
+    
+
+    def get_current_loss_scale(self) -> Array:
+        """Helper function to get current loss scale value for tuning dynamic loss scaling."""
+        if hasattr(self.loss_scaler, 'loss_scale'):
```
Comment (codekansas) [inline xax/task/mixins/mixed_precision.py:204]: when will this fail?

File: xax/task/mixins/mixed_precision.py (hunk 2 above)
Comment (ycheng223) [reply xax/task/mixins/mixed_precision.py:74]: Makes sense, better to end the run then have it run on a default setting and risk testing on the wrong precision.

I've updated the logic in those lines to now raise an exception:
image

Also updated mnist_mp.py to test the new exception logic (in def main at the bottom).
image
